File: fathersDay/Dsa/Day 29 (450. Delete Node in a BST)/Solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def deleteNode(self, root: 'TreeNode', key: int) -> 'TreeNode':
        # Base case: if root is None, return None
        if not root:
            return None
        if key < root.val:
            root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        else:
            # Node with only one child or no child
            if not root.left:
                return root.right
            elif not root.right:
                return root.left
        
            min_larger_node = self.getMin(root.right)
            root.val = min_larger_node.val
            root.right = self.deleteNode(root.right, min_larger_node.val)
        return root
    def getMin(self, node: 'TreeNode') -> 'TreeNode':
        while node.left:
            node = node.left
        return node
solution=Solution()
# Example usage:
# Constructing a binary search tree
root = TreeNode(5)
root.left = TreeNode(3)
root.right = TreeNode(6)
root.left.left = TreeNode(2)
root.left.right = TreeNode(4)
root.right.right = TreeNode(7)
key = 3
result = solution.deleteNode(root, key)
# Function to get the minimum value node in a subtree
print(result.val)  # Output: 5, as the node with value 3 is deleted and replaced by the minimum larger node (4)

File: fathersDay/Dsa/Day 29 (450. Delete Node in a BST)/Solution.cpp
#include <iostream>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (!root) return nullptr;

        if (key < root->val) {
            root->left = deleteNode(root->left, key);
        } else if (key > root->val) {
            root->right = deleteNode(root->right, key);
        } else {
            if (!root->left) {
                TreeNode* temp = root->right;
                delete root;
                return temp;
            } else if (!root->right) {
                TreeNode* temp = root->left;
                delete root;
                return temp;
            }
            TreeNode* min_larger_node = getMin(root->right);
            root->val = min_larger_node->val;
            root->right = deleteNode(root->right, min_larger_node->val);
        }
        return root;
    }

private:
    TreeNode* getMin(TreeNode* node) {
        while (node && node->left) {
            node = node->left;
        }
        return node;
    }
};

void print_inorder(TreeNode* node) {
    if (!node) return;
    print_inorder(node->left);
    std::cout << node->val << " ";
    print_inorder(node->right);
}

int main() {
    Solution solution;
    TreeNode* root = new TreeNode(5);
    root->left = new TreeNode(3);
    root->right = new TreeNode(6);
    root->left->left = new TreeNode(2);
    root->left->right = new TreeNode(4);
    root->right->right = new TreeNode(7);
    int key = 3;
    root = solution.deleteNode(root, key);
    print_inorder(root);
    std::cout << std::endl;
    return 0;
}
